contrib/largeNbDicts/largeNbDicts.c

```diff
@@ -397,6 +397,37 @@ typedef struct {
     size_t nbDDict;
 } ddict_collection_t;
 
+typedef struct {
+    ZSTD_CDict** cdicts;
+    size_t nbCDict;
+} cdict_collection_t;
+
+static const cdict_collection_t kNullCDictCollection = { NULL, 0 };
+
+static void freeCDictCollection(cdict_collection_t cdictc)
+{
+    for (size_t dictNb=0; dictNb < cdictc.nbCDict; dictNb++) {
+        ZSTD_freeCDict(cdictc.cdicts[dictNb]);
+    }
+    free(cdictc.cdicts);
+}
+
+/* returns .buffers=NULL if operation fails */
+static cdict_collection_t createCDictCollection(const void* dictBuffer, size_t dictSize, size_t nbCDict, int cLevel)
+{
+    ZSTD_CDict** const cdicts = malloc(nbCDict * sizeof(ZSTD_CDict*));
+    assert(cdicts != NULL);
```

Comment:
`CONTROL()` here and below

Reply:
That's a subtle one :
`CONTROL()` would be necessary if the case `cdicts==NULL` wasn't handled afterwards.

But it is.
In which case, I guess even the `assert()` is superfluous.

Reply:
Yeah, but it should be a control, since we `CONTROL()` for dictionary creation failures. It seems strange to sometimes return `NULL` and sometimes crash depending on which malloc fails. We don't gain anything from returning `NULL` from the first one.

Reply:
Yeah, right, so this is a matter of consistency now.

It depends on the end goal, do we want to lean towards a `CONTROL()` free code, which would require to chase down those `CONTROL()`, eventually in multiple rounds, and progressively replace them with proper return code, followed by appropriate error detection code at caller's side ?

Such a goal would be a necessity in a library code like `libzstd`.
But here, this is just a test tool.
So it doesn't seem really necessary.

Using `CONTROL()`simplifies the logic, because the caller can assume that the function never fails.

So, if I understand your point correctly, you believe that for consistency sake, rather than using `CONTROL()` as a "backup case", whenever an error case is not handled, the code should actively make use of it, and always use `CONTROL()` whenever applicable (there are still error cases that we want to detect and report, so it's not about pushing `CONTROL()` mindlessly everywhere).

Reply:
Yeah, I mean that this function both returns `NULL` and uses `CONTROL()`. That doesn't really make sense to me. Any particular function should either return an error code or use `CONTROL()`.

When a function uses both, we don't get the benefit of saying "we never crash", and we also have the burden of checking the return code, because not all conditions are `CONTROL()`ed.